Add muddiest points for RNA-seq part 1 (#29)
AlexsLemonade/2021-march-training: supplemental_notebooks/2021-03-23_muddiest_points.Rmd

It often requires a few more steps, but ultimately gives us more information and more control over our plot. 

We'll use the same `gastric-cancer` dataset we used in lecture but perform PCA outside of DESeq2 and show how to obtain gene loadings. 

![](diagrams/rna-seq_6.png)

These first steps are identical to what we did to set up this dataset in `02-gastric_cancer_exploratory-live.Rmd`. 

### Declare directories and files

```{r input-files}
# Main data directory
data_dir <- file.path("data", "gastric-cancer")

# directory with the tximeta processed data
txi_dir <- file.path(data_dir, "txi")
txi_file <- file.path(txi_dir, "gastric-cancer_tximeta.RDS")
```

```{r read-rds}
# Read in the RDS file we created in the last notebook
gene_summarized <- readr::read_rds(txi_file)
```

### Set up DESeq2 object

```{r ddset}
ddset <- DESeqDataSet(gene_summarized,
                      design = ~ tissue)
```

### Variance stabilizing transformation

```{r vst}
vst_data <- vst(ddset)
```

### Set up a matrix 

Now, we'll briefly show you how to use built-in functions for PCA. 
The PCA functions we are going to use can be used on any matrix.

> But because our data is a specialized object (a `SummarizedExperiment`), we will first extract the assay data so it is a matrix.

Reviewer: ```suggestion
But because our data is a specialized object (a`DESeqTransform`, which is structured similarly to the `SummarizedExperiment` we have used before), we will first extract the assay data as a matrix.
```

```{r}
vst_mat <- assay(vst_data)
```

Let's take a look at what the these data look like. 
Samples are columns and rows are genes. 

```{r}
head(vst_mat)
```

> `prcomp()` expects the features (in our case, genes) to be columns, but in our data and often gene matrices in general, genes are actually rows. 

Reviewer 2: ```suggestion
`prcomp()` expects the features (in our case, genes) to be columns, but in our data and often gene expression matrices in general, genes are actually rows. 
```

So we have to transpose our data (make the rows columns and the columns rows).

```{r}

> # The function t() transposes data.frames or matrices and spits out a matrix

Reviewer: ```suggestion
# The function t() transposes data.frames or matrices and returns a matrix
```

vst_mat_transposed <- t(vst_mat)

# Preview first six columns of newly transposed data
head(vst_mat_transposed[, 1:6])
```

### Principal component analysis using prcomp()

Now that our matrix is set up, we are ready to run `prcomp()`.

> Often you will want to use the argument `scale = TRUE` but for Variance Stabilizing Transformed data, this is not necessary. 

Author: This is what I don't know for sure is true.

Reviewer 2: A few things:

- I think it's fine not to use `scale = TRUE` in this case, and in fact, looking at [the code for the `plotPCA()` wrapper function](https://github.com/mikelove/DESeq2/blob/5a32b40b563e7db35fc3dafed705269695ae2dca/R/plots.R#L217), there is no scaling involved.
- However, there's not enough information about what `scale = TRUE` _does_ to say "often you will want to use it."
- You might want to subset to high variance genes (whatever definition we use; `DESeq2` uses the top 500), which would be instructive and allow you to compare and contrast with and without `scale = TRUE` if you'd like (e.g., you wouldn't have the problem with zero variance features).

Reviewer: Honestly, I often don't know whether scaling is recommended or not. It really depends on whether you want to weight the results by variance of the individual elements. Sometimes you do, sometimes you really don't. My experience with gene expression PCA is that it usually doesn't matter too much, but that is a pretty rough sense, not empirical.

In this case, I would say you definitely will get good results from subsetting to the most variable genes, and I would recommend that. At the very least, you can and should remove genes with no variance. They will not contribute to the results at all, and just mean you take more memory.

Author: Will subset to high variance genes then!

```{r}
pca_results <- prcomp(vst_mat_transposed)
```

What does `prcomp()` output look like? It's a list with named items.
For named items, you can extract them with `$` as the output from `str()` suggests. 

```{r}
str(pca_results)
```

We'll take a look at a couple of the items in `pca_results`. 

> `$x` holds the principal components themselves (PC1, PC2, etc).

Reviewer: You probably want to say just a bit more here. It is a matrix, what are the rows and columns?

Although it can be helpful, they say essentially: 

> A good rule of thumb may be to treat any interpretations about loading as a hypothesis that needs to be validated by a completely independent means. This is definitely one of those areas where you want to be conservative.

So whatever you may find with your gene loadings, you may want to do some further evaluations on before resting on any conclusions. 
This is in part because this can be a precarious endeavor if your data is noisy or has not been normalized properly. 

That being said, we'll take a look at how we can see this information. 
For example, what's the top 10 genes correlated to PC1? 
First, we can set up a data frame with this information. 

```{r}
pc_loadings_df <- data.frame(pca_results$rotation) %>% 
  tibble::rownames_to_column("gene_id") %>% 
  dplyr::top_n(PC1, n = 10)
```

If you wanted to see the genes with negative loadings, you would need to use the absolute values sort using the `bottom_n()` function. 

Now we can plot these top results.

```{r}
ggplot(pc_loadings_df, aes(gene_id, PC1)) + 
  geom_bar(stat = "identity") + 
  theme_classic() +

>   theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) 

Reviewer: probably worth a comment here

Author: Do we even want this plot anyway though?

```

Do we prefer gene symbols instead? Our tximeta has this info for us. 

```{r}
annot_pc_loadings_df <- pc_loadings_df %>% 
  dplyr::left_join(data.frame(rowData(vst_data)))

ggplot(annot_pc_loadings_df, aes(reorder(symbol, -PC1), PC1)) + 
  geom_bar(stat = "identity") + 
  theme_classic() +
  theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) + 
  xlab("gene symbol")
```

> ## Finding proportion variance with prcomp() output 

Reviewer: I would report this before getting into the loadings. I would also probably drop the plotting of the loadings and instead plot higher order PCs, as I think that is more useful, and a real advantage of `prcomp()` over plotPCA which only every gets the top two.

It can be useful to understand the proportion of variance explained by each principal component when visualizing and interpreting the results.
For example, if PC1 explained 96% of the variance in your data and very clearly showed a difference between sample batches you would be very concerned!
On the other hand, if a separation of batches was apparent in a different principal component that explained a low proportion of variance and the first few PCs explained most of the variance and appeared to correspond to something like tissue type and treatment, you would be less concerned.

`summary()` will report the proportion of variance explained by each principal component.
By accessing the `importance` element with `<summary results>$importance`, we can use indexing to only look at the first 10 PCs.

```{r pca_summary}
# Save summary of the PCA results
pca_summary <- summary(pca_results)

# Importance information for the first 10 PCs

> pca_importance <- pca_summary$importance[, 1:6]

Reviewer: First 10 or first 6?

Author: Copied from elsewhere so I forgot to update comment.

---
title: "Day 2: RNA-seq Supplementary Notebook"
author: Candace Savonen
date: 2021-03-23
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

## Introduction

This notebook is designed to address questions that came up in the Introduction to RNA-seq module during CCDL training. 

- [A bit on file paths and file organization](#file-paths)
- [How to find out how much each gene contributes to a particular Principal Component and other fun things with PCA](#principal-components-analysis)

> 

Reviewer: We should probably add a little note on how to download this notebook and put it in the correct place so that the paths all work.

# File Paths

> ![The directory structure of this particular project](https://raw.githubusercontent.com/AlexsLemonade/training-modules/master/RNA-seq/diagrams/gastric_cancer_structure.png)

Reviewer: For stability.
```suggestion
![The directory structure of this particular project](https://raw.githubusercontent.com/AlexsLemonade/training-modules/2021-march/RNA-seq/diagrams/gastric_cancer_structure.png)
```

- [Quantitative Understanding in Biology Principal Component Analysis](https://physiology.med.cornell.edu/people/banfelder/qbio/lecture_notes/3.4_Principal_component_analysis.pdf)

Instead of using DESeq2 to run our PCA and plotting for us, we can do this using `prcomp()` function and ggplot2.
It often requires a few more steps, but ultimately gives us more information and more control over our plot. 

We'll use the same `gastric-cancer` dataset we used in lecture but perform PCA outside of DESeq2 and show how to obtain gene loadings. 

![](diagrams/rna-seq_6.png)

These first steps are identical to what we did to set up this dataset in `02-gastric_cancer_exploratory-live.Rmd`. 

### Declare directories and files

```{r input-files}
# Main data directory
data_dir <- file.path("data", "gastric-cancer")

# directory with the tximeta processed data
txi_dir <- file.path(data_dir, "txi")
txi_file <- file.path(txi_dir, "gastric-cancer_tximeta.RDS")
```

```{r read-rds}
# Read in the RDS file we created in the last notebook
gene_summarized <- readr::read_rds(txi_file)
```

### Set up DESeq2 object

```{r ddset}

> ddset <- DESeqDataSet(gene_summarized,

Reviewer: DESeq2 is not explicitly loaded. We should add that.

                      design = ~ tissue)
```

### Variance stabilizing transformation

```{r vst}
vst_data <- vst(ddset)
```

### Set up a matrix 

Now, we'll briefly show you how to use built-in functions for PCA. 
The PCA functions we are going to use can be used on any matrix.

But because our data is a specialized object (a`DESeqTransform`, which is structured similarly to the `SummarizedExperiment` we have used before), we will first extract the assay data as a matrix.

```{r}
vst_mat <- assay(vst_data)
```

> Often PCA will be more informative if we only run it on a high variance subset of the genes. 

Reviewer: ```suggestion
Often PCA will be as informative (and faster to calculate) if we only run it on a high variance subset of the genes. 
```

> So we will use the top 500 genes in this example. 

Reviewer: ```suggestion
Following the default behavior of the `DESeq2` function `plotPCA()`that we used previously, we will use the top 500 genes in this example. 
```

Here, `apply(., 1, var)` calculates the variance for each row/gene. 

```{r}
high_var_subset <- vst_mat %>% 
  as.data.frame() %>% 
  dplyr::top_n(apply(., 1, var), n = 500)
```

Let's take a look at what the these data look like. 
Samples are columns and rows are genes. 

```{r}
head(high_var_subset)
```

> `prcomp()` expects the features (in our case, genes) to be columns, but in our data and often gene expression matrices in general, genes are actually rows. 

Reviewer: ```suggestion
`prcomp()` expects the features (in our case, genes) to be columns and the observations (samples) to be rows, but in our data and often gene expression matrices in general, genes are actually rows. 
```

So we have to transpose our data (make the rows columns and the columns rows).

```{r}

> # The function t() transposes data.frames or matrices and returns a matrix

Reviewer: ```suggestion
# The function t() transposes data frames or matrices and returns a matrix
```

high_var_transposed <- t(high_var_subset)

# Preview first six columns of newly transposed data
head(high_var_transposed[, 1:6])
```

### Principal component analysis using prcomp()

Now that our matrix is set up, we are ready to run `prcomp()`.

> Sometimes you will want to use the argument `scale = TRUE` but for Variance Stabilizing Transformed data, this is not necessary. 

Reviewer: If we aren't going to explain why, I think we can delete this sentence.

```{r}
pca_results <- prcomp(high_var_transposed)
```

What does `prcomp()` output look like? It's a list with named items.
For named items, you can extract them with `$` as the output from `str()` suggests. 

```{r}
str(pca_results)
```

We'll take a look at a couple of the items in `pca_results`. 

`$x` holds the principal components themselves (PC1, PC2, PC3 etc).

> The data in `$x` is a matrix where rows are samples and columns are the components.  
> You will have as many components as you do samples. 

Reviewer: ```suggestion
The data in `$x` is a matrix where rows are samples and columns are the principal components.
You will have as many principal components as you do samples (or really whichever dimension of your matrix is smaller, but that is almost always samples!). 
```

```{r}
# The principal component representation is returned in x
pca_results$x[1:6, 1:6]
```

`$rotation` holds the loadings -- aka how much does each feature/gene contribute to each principal component?
Loadings can be interpreted much like correlation coefficients: they can be positive or negative and can differ in magnitude. 

```{r}
# The loadings are returned in $rotation
pca_results$rotation[1:6, 1:6]
```

### Plot PC1 and PC2

> Let's plot the first PC components as we have elsewhere. 

Reviewer: ```suggestion
Let's plot the first principal components as we have elsewhere. 
```

First we will set up a data frame for plotting and join the `colData` from `vst_data`. 

```{r}
pca_df <- data.frame(pca_results$x) %>% 
  tibble::rownames_to_column("names") %>% 
  dplyr::inner_join(data.frame(colData(vst_data)))
```

Let's make a scatter plot of PC1 and PC2.  
It will be nearly identical to the PCA plot we made in `02-gastric_cancer_exploratory.Rmd`. 

```{r}

> ggplot(pca_df, aes(PC1, 

Reviewer: I don't think we ever explicitly loaded ggplot2 in this notebook, probably worth doing here.

Author: I had it loaded (and the other packages) but in my re-arrangments it got lost. Whoops.

                   PC2, 
                   color = tissue)) + 
  geom_point(size = 3) + 
  theme_classic()
```

With some minor changes to the plot above, we can also plot the other PCs like PC3 and PC4.  
This is an advantage we have with having the data from `prcomp()` -- we have all the PCs available to us. 

```{r}
ggplot(pca_df, aes(PC3, # Change these to PC3 and PC4
                   PC4, 
                   color = tissue)) + 
  geom_point(size = 3) + 
  theme_classic()
```

> ## Finding proportion variance with prcomp() output 

Reviewer: ```suggestion
## Finding the proportion of variance explained with `prcomp()` output 
```

For example, if you sets of samples that use different RNA selection strategies, looking at the loadings for PC1 may be helpful. 
If in a poly-A selected sample there high loadings for genes that are not poly-adenylated (e.g., a bunch of histones) or if a ribo-minus selected sample
has ribosomal RNA show up as the top hits this may indicate that RNA selection methods didn't thoroughly process. 

That being said, we'll take a look at how we can see this information. 
For example, what's the top 6 genes correlated to PC1? 
First, we can set up a data frame with this information. 

```{r}
pc_loadings_df <- data.frame(pca_results$rotation) %>% 
  tibble::rownames_to_column("gene_id") %>% 
  dplyr::top_n(PC1, n = 6)
```

If you wanted to see the genes with negative loadings, you would need to use the absolute values sort using the `bottom_n()` function. 

Now we can plot these top results.

```{r}
ggplot(pc_loadings_df, aes(gene_id, PC1)) + 
  geom_bar(stat = "identity") + 
  theme_classic() +
  # This theme function and arguments are adjusting the labels so they are easier to read. 

>   theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) 

Reviewer: ```suggestion
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
```

```

Do we prefer gene symbols instead? Our tximeta has this info for us. 

```{r}
annot_pc_loadings_df <- pc_loadings_df %>% 
  dplyr::left_join(data.frame(rowData(vst_data)))

ggplot(annot_pc_loadings_df, aes(reorder(symbol, -PC1), PC1)) + 
  geom_bar(stat = "identity") + 
  theme_classic() +
  # This theme function and arguments are adjusting the labels so they are easier to read. 

>   theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) + 

Reviewer: ```suggestion
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
```

---
title: "Day 2: RNA-seq Supplementary Notebook"
author: Candace Savonen
date: 2021-03-23
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

## Introduction

This notebook is designed to address questions that came up in the Introduction to RNA-seq module during CCDL training. 

- [A bit on file paths and file organization](#file-paths)
- [How to find out how much each gene contributes to a particular Principal Component and other fun things with PCA](#principal-components-analysis)

### How to use this notebook: 

> 1) Click the `download` button to put this Rmd (`2021-03-23-_muddiest_points.Rmd`) on your computer.   

Reviewer: ```suggestion
1) Click the `Code` button in the upper right of this page and choose `Download Rmd` to save this notebook (`2021-03-23_muddiest_points.Rmd`) to your computer.   
```

may be very sensitive to noise in the data...

> A good rule of thumb may be to treat any interpretations about loading as a hypothesis that needs to be validated by a completely independent means. This is definitely one of those areas where you want to be conservative.

So in most cases, you may want to avoid this kind of analysis unless you have a particular hypothesis that you are looking to check.
Even when you do look at this kind of analysis, you may want to do some further evaluations on before resting on any conclusions. 
This is in part because this can be a precarious endeavor and definitely cannot be trusted if your data is noisy or has not been normalized properly. 

For example, if you sets of samples that use different RNA selection strategies, looking at the loadings for PC1 may be helpful. 
If in a poly-A selected sample there high loadings for genes that are not poly-adenylated (e.g., a bunch of histones) or if a ribo-minus selected sample
has ribosomal RNA show up as the top hits this may indicate that RNA selection methods didn't thoroughly process. 

That being said, we'll take a look at how we can see this information. 

> For example, what's the top 6 genes correlated to PC1? 

Reviewer: ```suggestion
For example, what are the top 6 genes that contribute to PC1? 
```

> First, we can set up a data frame with this information. 

Reviewer: ```suggestion
First, we can set up a data frame with the loading (rotation) information. 
```

```{r}
pc_loadings_df <- data.frame(pca_results$rotation) %>% 
  tibble::rownames_to_column("gene_id") %>% 
  dplyr::top_n(PC1, n = 6)
```

> If you wanted to see the genes with negative loadings, you would need to use the absolute values sort using the `bottom_n()` function. 

Reviewer: ```suggestion
If you wanted to see the genes with negative loadings, you would need to use sort using the `bottom_n()` function or use the absolute values of the loadings. 
```

has ribosomal RNA show up as the top hits this may indicate that RNA selection methods didn't thoroughly process. 

That being said, we'll take a look at how we can see this information. 
For example, what are the top 6 genes that contribute to PC1? 
First, we can set up a data frame with the loading (rotation) information. 

```{r}
pc_loadings_df <- data.frame(pca_results$rotation) %>% 
  tibble::rownames_to_column("gene_id") %>% 
  dplyr::top_n(PC1, n = 6)
```

If you wanted to see the genes with negative loadings, you would need to use sort using the `bottom_n()` function or use the absolute values of the loadings. 

Now we can plot these top results.

```{r}
ggplot(pc_loadings_df, aes(gene_id, PC1)) + 
  geom_bar(stat = "identity") + 
  theme_classic() +
  # This theme function and arguments are adjusting the labels so they are easier to read. 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
```

Do we prefer gene symbols instead? Our tximeta has this info for us. 

```{r}
annot_pc_loadings_df <- pc_loadings_df %>% 
  dplyr::left_join(data.frame(rowData(vst_data)))

> ggplot(annot_pc_loadings_df, aes(reorder(symbol, -PC1), PC1)) + 

Reviewer: The reordering is going to confuse people if it isn't explained. If you are going to do it at all, you should do it in both plots and explain it, or not do it at all and mention (which you should do either way) that the default order is going to be alphabetical by symbol so it will change between these two plots.

I am nbot sure we need the second plot at all, tbh.

Author: I thought that's what you said you wanted in a previously comment: https://github.com/AlexsLemonade/2021-march-training/pull/29#discussion_r600430875

Reviewer: I don't see how that comment is related to this section? I mean specifically the second plot in this section (by symbol rather than ensembl id)
